testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java

```diff
@@ -0,0 +1,316 @@
+        /**
+         * Configure retry-queueing feature for the subscription
+         */
+        private RetryConfig retryConfig;
+        /**
+         * Supply additional {@link ProcessorProperties} through {@link PropertySupplier}
+         */
+        private PropertySupplier propertySupplier;
+        /**
+         * Exclude semantics from assertion.
+         * Intended to be used when we test a feature which breaks subset of semantics
+         */
+        public Builder excludeSemantics(GuaranteeType... guarantees) {
+            for (GuaranteeType guarantee : guarantees) {
+                defaultSemantics.remove(guarantee);
+            }
+            return this;
+        }
+        /**
+         * Include additional semantics in assertion.
+         * Feature-specific processing guarantee will be injected through this method
+         */
+        public Builder customSemantics(ProcessingGuarantee... guarantees) {
+            customSemantics.addAll(Arrays.asList(guarantees));
+            return this;
+        }
+
+        private Builder(KafkaClusterRule rule) {
+            this.rule = rule;
+        }
+
+        public ProcessorTestSuite build() {
+            Set<ProcessingGuarantee> semantics = new HashSet<>();
+            for (GuaranteeType guaranteeType : defaultSemantics) {
+                semantics.add(guaranteeType.get());
+            }
+            semantics.addAll(customSemantics);
+
+            return new ProcessorTestSuite(rule,
+                                          configureProcessorsBuilder,
+                                          retryConfig,
+                                          propertySupplier,
+                                          semantics);
+        }
+    }
+
+    public static Builder builder(KafkaClusterRule rule) {
+        return new Builder(rule);
+    }
+
+    /**
+     * Run the test.
+     *
+     * Can be called only once per {@link ProcessorTestSuite} instance since
+     * running the test possibly mutates {@link ProcessingGuarantee}'s internal state
+     */
+    public void run() {
+        String topic = rule.admin().createRandomTopic(NUM_PARTITIONS, 3);
+        CountDownLatch rollingRestartLatch = new CountDownLatch(NUM_TASKS / 2);
+        Producer<String, DecatonTaskRequest> producer = null;
+        ProcessorSubscription[] subscriptions = new ProcessorSubscription[NUM_SUBSCRIPTION_INSTANCES];
+
+        try {
+            producer = TestUtils.producer(rule.bootstrapServers());
+            for (int i = 0; i < subscriptions.length; i++) {
+                subscriptions[i] = newSubscription(i, topic, Optional.of(rollingRestartLatch));
+            }
+            CompletableFuture<Map<Integer, Long>> produceFuture =
+                    produceTasks(producer, topic, record -> semantics.forEach(g -> g.onProduce(record)));
+
+            performRollingRestart(subscriptions, i -> newSubscription(i, topic, Optional.empty()));
```

Comment:
`rollingRestartLatch.await()` is missing here?

Reply:
oops nice catch